library/core/src/ffi/c_str.rs

```diff
@@ -196,15 +196,21 @@ impl CStr {
     /// allows inspection and interoperation of non-owned C strings. The total
     /// size of the raw C string must be smaller than `isize::MAX` **bytes**
     /// in memory due to calling the `slice::from_raw_parts` function.
-    /// This method is unsafe for a number of reasons:
     ///
-    /// * There is no guarantee to the validity of `ptr`.
-    /// * The returned lifetime is not guaranteed to be the actual lifetime of
-    ///   `ptr`.
-    /// * There is no guarantee that the memory pointed to by `ptr` contains a
-    ///   valid nul terminator byte at the end of the string.
-    /// * It is not guaranteed that the memory pointed by `ptr` won't change
-    ///   before the `CStr` has been destroyed.
+    /// # Safety
+    ///
+    /// * `data` must be [valid] for reads for `len` many bytes,
+    ///   and it must be properly aligned. This means in particular:
+    ///
+    ///     * The entire memory range of this slice must be contained within a single allocated object!
+    ///     * `data` must be non-null and aligned even for a zero-length cstr.
+    ///
+    /// * The caller must ensure that the returned lifetime represents the actual
+    ///   lifetime of `ptr`.
+    /// * The memory pointed to by ptr must contain a valid nul terminator a the
+    ///   end of the string.
```

Comment:
```suggestion
    /// * The memory pointed to by `ptr` must contain a valid nul terminator at the
    ///   end of the string.
```

Reply:
This should probably go first; you can also use it to define `len` which you will need later.